### PR TITLE
Chore: Migrate yarn from v1 to v2

### LINKED ISSUE #23018: @grafana/ui: To add new package, yarn needs to be downgraded to 1.18
**What happened**:
When adding new packages to @grafana/ui, following error is thrown:`An unexpected error occurred: "expected workspace package to exist for ...`. Therefore it is not possible to add any new packages.

Downgrading yarn to v 1.18 fixes the issue. More info: https://github.com/yarnpkg/yarn/issues/7807.

**Environment**:
- Grafana version: master
- OS Grafana is installed on: OS X

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Migrates the repo from yarn v1 to yarn v2. [Here](https://yarnpkg.com/getting-started/qa#why-should-you-upgrade-to-yarn-modern) are some reasons why we might want to do that.
Note that this doesn't make use of the "Plug'n'Play" strategy, but that is something that could be done in the future.

**Special notes for your reviewer**:
I started fixing some of the many peer dependency issues reported by yarn before deciding that that might be better suited for a future PR, so that's what some of the dependency additions are about.

Fixes https://github.com/grafana/grafana/issues/23018

### CODE REVIEW

**Comment by dprokop** [comment]
image

This is scary :D Think you are about to break some records @kaydelaney!

**Comment by kaydelaney** [comment]
At @jackw's recommendation I think we'll be doing this upgrade in a few steps:
1. Switch to yarn 2 with `nodeLinker: node-modules`
2. Make a separate PR for introducing [zero installs](https://yarnpkg.com/features/zero-installs/)
3. Fix (peer) dependency issues and switch to `nodeLinker: pnp`

**Comment by kaydelaney** [comment]
So here's something I'd like some second opinions on: Yarn 2 does away with "nohoist" which we currently use to stop types from being hoisted, which (in combination with some tsconfig settings) ensures we aren't referring to any dependencies outside of the given workspace, without completely preventing hoisting and duplicating a bunch of packages.
What yarn 2 has instead of "nohoist" is what they call "hoistingLimits", which you can set to one of three options.
The default ("none") just hoists everything as high as possible. Then there's "workspaces" which doesn't hoist anything outside of the workspace that depends on it, and finally there's "dependencies" which doesn't hoist anything past the direct dependencies for each workspace.
So here's the problem, we _could_ go with "workspaces", which would allow us to keep that behavior of not accidentally using dependencies outside of a workspace, but since these new hoisting limits don't offer the granularity of "nohoist", we'd have a lot of duplicate packages. I did a quick `du` comparison and it looks like it increases the overall size of the repo by more than a gigabyte, so not great.
The other option then is just to not use any hoisting limits at all, but of course then there's no safeguard against using dependencies outside of a workspace.
So both options have a downside, and I'm not really sure which would be better to be honest. What do you think?

Note: I think(?) the pnp linker solves these problems but I really have no estimate on how much effort it would take to migrate to that.

File: package.json [hunk 1]
```diff
@@ -206,28 +208,39 @@
     "sass-lint": "1.12.1",
     "sass-loader": "12.1.0",
     "sinon": "8.1.1",
+    "sockjs-client": "1.5.2",
```
Comment (ryantxu) [inline package.json:211]: Why sockjs?  We used to have it for live, but not used anymore

File: package.json [hunk 2]
```diff
@@ -175,7 +177,7 @@
     "html-webpack-harddisk-plugin": "2.0.0",
     "html-webpack-plugin": "5.3.2",
     "husky": "4.2.1",
-    "iconscout-unicons-tarball": "https://github.com/grafana/icons/tarball/9728be621a4e7d891611149c9cd179e793f790a7",
+    "iconscout-unicons-tarball": "https://github.com/grafana/icons/tarball/63056cd833ba7ee4e94904492b3a8c0cabc38d28",
```
Comment (ryantxu) [inline package.json:180]: Is this change required for the yarn change?

Comment (kaydelaney) [reply package.json:180]: Yes. The previous commit doesn't include a package.json manifest in the project root, which is required with yarn 2.

**Comment by kaydelaney** [comment]
I think I've found a good solution to the problem outlined in my last comment. Specifying `nmMode: hardlinks-global` (or `hardlinks-local`, but global saves a bit more space) in the yarn config saves multiple gigabytes of space to the point that the total size of the repo (after dependencies are installed) falls below what we have now, even.
So I think the plan will be to go with that and `hoistingLimits: workspaces`.

**Comment by merceyz** [comment]
The issue you're running into was fixed in https://github.com/yarnpkg/berry/pull/3204 which is currently in the canary release
```
yarn set version canary
```

**Comment by merceyz** [comment]
I don't know if it will fix the issues you're running into but the easiest way to deal with ESLint and its resolution has been to apply a patch that changes it to be correct.
https://github.com/microsoft/rushstack/tree/1bafd7cd2cc3b5344149a2b78317efd6a362e31f/stack/eslint-patch

**Comment by kaydelaney** [comment]
> I don't know if it will fix the issues you're running into but the easiest way to deal with ESLint and its resolution has been to apply a patch that changes it to be correct.
> https://github.com/microsoft/rushstack/tree/1bafd7cd2cc3b5344149a2b78317efd6a362e31f/stack/eslint-patch

Oh this looks really interesting, will definitely take a look!

File: package.json [hunk 3]
```diff
@@ -323,22 +336,24 @@
...
     "uuid": "8.3.0",
     "visjs-network": "4.25.0",
     "whatwg-fetch": "3.1.0"
   },
   "resolutions": {
     "caniuse-db": "1.0.30000772",
-    "underscore": "1.12.1"
+    "underscore": "1.12.1",
+    "@types/slate": "0.47.2",
+    "@microsoft/api-extractor-model": "7.10.3",
```
Comment (kaydelaney) [inline package.json:348]: These dependency resolutions have been added to fix how they've been listed for our forks of api-extractor and api-documenter. When built/distributed, those packages should have had any workspace ranges transformed to the appropriate version, but for whatever reason in the distributed package.json the versions are still listed as "workspace:*", which causes a problem for yarn 2. Hopefully we can fix the actual issue with the forks at some point, but for now this solves the problem.

File: packages/grafana-ui/package.json [hunk 4]
```diff
@@ -56,23 +55,21 @@
...
-    "react-dom": "17.0.1",
     "react-dropzone": "11.3.4",
     "react-highlight-words": "0.17.0",
-    "react-hook-form": "7.5.3",
     "react-inlinesvg": "2.3.0",
     "react-popper": "2.2.4",
-    "react-router-dom": "^5.2.0",
     "react-select": "4.3.0",
     "react-table": "7.7.0",
+    "react-test-renderer": "17.0.1",
```
Comment (zoltanbedi) [inline packages/grafana-ui/package.json:68]: Why is this a dependency? I assume this is only needed for testing.

Comment (kaydelaney) [reply packages/grafana-ui/package.json:68]: You're right, will fix

**Comment by merceyz** [comment]
I see you're caching the `node_modules` folders, you should change that to cache the Yarn cache itself which is much more efficient.
https://github.com/actions/cache/blob/6bbe742add91b3db4abf110e742a967ec789958f/examples.md#node---yarn-2
You should remove `${{ runner.os }}-` from the key as that is unnecessary

**Comment by kaydelaney** [comment]
> I see you're caching the `node_modules` folders, you should change that to cache the Yarn cache itself which is much more efficient.
> https://github.com/actions/cache/blob/6bbe742add91b3db4abf110e742a967ec789958f/examples.md#node---yarn-2
> You should remove `${{ runner.os }}-` from the key as that is unnecessary

Nice catch! Made the change, but let me know if I've done something wrong (not too much experience with github actions).

File: packages/grafana-ui/package.json [hunk 5]
```diff
@@ -123,10 +120,27 @@
     "rollup-plugin-typescript2": "0.29.0",
     "rollup-plugin-visualizer": "4.2.0",
     "storybook-dark-mode": "1.0.8",
-    "ts-loader": "8.0.11",
     "typescript": "4.4.3",
     "webpack-filter-warnings-plugin": "1.2.1"
   },
+  "peerDependencies": {
```
Comment (jackw) [inline packages/grafana-ui/package.json:126]: Do we know what sort of impact (if any) these packages moving into `peerDependencies` will have on plugin devs?

File: packages/grafana-toolkit/package.json [hunk 6]
```diff
@@ -58,10 +57,6 @@
     "copy-webpack-plugin": "5.1.2",
     "css-loader": "3.4.2",
     "eslint": "7.21.0",
-    "eslint-config-prettier": "7.2.0",
```
Comment (jackw) [inline packages/grafana-toolkit/package.json:61]: Do we know what sort of impact (if any) deleting these packages will have on plugin devs?

**Review by jackw — COMMENTED** [review]
Absolutely amazing work on this @kaydelaney and @hugohaggmark! 🥇👏

File: .github/workflows/publish.yml [hunk 7]
```diff
@@ -16,16 +16,19 @@ jobs:
     steps:
     - uses: actions/checkout@v1
     - run: git clone --single-branch --no-tags --depth 1 -b master https://grafanabot:${{ secrets.GH_BOT_ACCESS_TOKEN }}@github.com/grafana/website-sync ./.github/actions/website-sync
+    - name: Get yarn cache directory path
+      id: yarn-cache-dir-path
+      run: echo "::set-output name=dir::$(yarn config get cacheFolder)"
```
Comment (merceyz) [inline .github/workflows/publish.yml:21]: You should move the `actions/setup-node` step so it runs before this, `yarn` needs `node` to run

File: .github/workflows/publish.yml [hunk 8]
```diff
@@ -16,16 +16,19 @@ jobs:
...
     - uses: actions/cache@v2.1.6
       with:
-        path: '**/node_modules'
-        key: ${{ runner.os }}-modules-${{ hashFiles('**/yarn.lock') }}
+        path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+        key: yarn-${{ hashFiles('**/yarn.lock') }}
```
Comment (merceyz) [inline .github/workflows/publish.yml:25]:
```suggestion
        key: yarn-${{ hashFiles('**/yarn.lock') }}
        restore-keys: |
            yarn-
```
This will allow you to reuse as much of the cache as possible when the lockfile changes

File: .github/workflows/publish.yml [hunk 9]
```diff
@@ -16,16 +16,19 @@ jobs:
...
     - name: generate-packages-docs
       uses: actions/setup-node@v2.4.0
       id: generate-docs
       with:
         node-version: '14'
-    - run: yarn install --pure-lockfile --no-progress
+    - run: YARN_ENABLE_PROGRESS_BARS=false yarn install --immutable
```
Comment (merceyz) [inline .github/workflows/publish.yml:31]: I see `YARN_ENABLE_PROGRESS_BARS=false` a lot in this PR, Yarn sets that automatically when it detects its running in a CI https://yarnpkg.com/configuration/yarnrc#enableProgressBars so it's unnecessary here, unsure if `drone` is detected as a CI though

File: .drone.yml [hunk 10]
```diff
@@ -440,7 +440,7 @@ steps:
   image: buildkite/puppeteer
   commands:
   - yarn wait-on http://$HOST:$PORT
-  - yarn -s test:accessibility --json > pa11y-ci-results.json
+  - yarn run --silent test:accessibility --json > pa11y-ci-results.json
```
Comment (merceyz) [inline .drone.yml:443]:
```suggestion
  - yarn run test:accessibility --json > pa11y-ci-results.json
```
The `--silent` flag doesn't do anything
https://github.com/yarnpkg/berry/blob/67e046fec96402637e367b39502e185c82696db5/packages/plugin-essentials/sources/commands/run.ts#L55-L60

File: package.json [hunk 11]
```diff
@@ -38,9 +38,9 @@
     "prestart": "yarn themes:generate",
     "prestart:hot": "yarn themes:generate",
     "prestart:noTsCheck": "yarn themes:generate",
-    "start": "webpack --progress --color --watch --env noTsCheck=0 --config scripts/webpack/webpack.dev.js",
-    "start:hot": "webpack serve --progress --color --config scripts/webpack/webpack.hot.js",
-    "start:noTsCheck": "webpack --progress --color --watch --env noTsCheck=1 --config scripts/webpack/webpack.dev.js",
+    "start": "yarn prestart && webpack --progress --color --watch --env noTsCheck=0 --config scripts/webpack/webpack.dev.js",
```
[the remark this answers is not in the document]
Comment (merceyz) [reply package.json:41]: It was indeed https://yarnpkg.com/getting-started/migration#explicitly-call-the-pre-and-post-scripts

[the remark this answers is not in the document]
Comment (kaydelaney) [reply package.json:41]: Makes sense, good idea!

File: packages/grafana-ui/src/components/Slider/_Slider.scss [hunk 12]
```diff
@@ -1 +1 @@
-@import '../../node_modules/rc-slider/assets/index.css';
+@import '../../packages/grafana-ui/node_modules/rc-slider/assets/index.css';
```
Comment (merceyz) [inline packages/grafana-ui/src/components/Slider/_Slider.scss:1]:
```suggestion
@import 'rc-slider/assets/index.css';
```
https://sass-lang.com/documentation/at-rules/import

File: package.json [hunk 13]
```diff
@@ -349,6 +366,7 @@
   },
   "volta": {
     "node": "14.16.0",
-    "yarn": "1.22.5"
-  }
+    "yarn": "3.1.0-rc.5"
+  },
+  "packageManager": "yarn@3.1.0-rc.5"
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply package.json:371]: https://github.com/yarnpkg/berry/pull/3204
Using `hardlinks-global` causes problems on CI with the stable build unfortunately. Fixed in canary which is why we're using it here.
As for the dockerfile, I'm not sure, but it's probably best to update that yarn version string. Good catch.

File: packages/grafana-ui/package.json (hunk 5 above)
Comment (kaydelaney) [reply packages/grafana-ui/package.json:126]: Not sure, though CI passes so I guess it's not completely broken at least. In any case I think plugins are the textbook usecase for `peerDependencies` so I would hope it wouldn't cause any issues. I'll investigate a bit more carefully though.

File: .drone.yml (hunk 10 above)
Comment (kaydelaney) [reply .drone.yml:443]: Oh nice, thanks

File: packages/grafana-ui/package.json (hunk 5 above)
Comment (jackw) [reply packages/grafana-ui/package.json]: I'm thinking that depending on the package manager the dev is using those dependencies may not be installed now they are peer dependencies. I guess we can follow up on it once this PR is merged and the packages are published to see what impact there is and get fixes in place before 8.3 and/or document breaking changes.

File: packages/grafana-toolkit/package.json (hunk 6 above)
Comment (kaydelaney) [reply packages/grafana-toolkit/package.json:61]: There shouldn't be any impact, as we added `@rushstack/eslint-patch` which lets eslint configs (like `@grafana/eslint-config`) use their own dependencies, which includes `eslint-config-prettier`.

**Review by jackw — APPROVED** [review]
Just taken this for a spin and seems to be working great! Amazing stuff @kaydelaney @hugohaggmark! 🚀

**Review by ashharrison90 — APPROVED** [review]
also had a play, works great! 👏👏👏👏👏

**Review by zoltanbedi — APPROVED** [review]
lgtm

File: .github/workflows/publish.yml [hunk 14]
```diff
@@ -16,16 +16,21 @@ jobs:
...
-        key: ${{ runner.os }}-modules-${{ hashFiles('**/yarn.lock') }}
+        path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+        key: yarn-${{ hashFiles('**/yarn.lock') }}
+        restore-keys: |
+            yarn-
     - name: generate-packages-docs
-      uses: actions/setup-node@v2.4.1
       id: generate-docs
       with:
         node-version: '14'
```
Comment (merceyz) [inline .github/workflows/publish.yml:32]: This is broken now, you only moved the `uses` part
